exonum/src/events/noise/sodium_resolver.rs

```diff
@@ -0,0 +1,243 @@
+use byteorder::{ByteOrder, LittleEndian};
+use rand::{thread_rng, Rng};
+use snow::params::{CipherChoice, DHChoice, HashChoice};
+use snow::types::{Cipher, Dh, Hash, Random};
+use snow::{CryptoResolver, DefaultResolver};
+
+use sodiumoxide::crypto::aead::chacha20poly1305 as sodium_chacha20poly1305;
+use sodiumoxide::crypto::hash::sha256 as sodium_sha256;
+use sodiumoxide::crypto::scalarmult::curve25519 as sodium_curve25519;
+
+pub struct SodiumResolver {
+    parent: DefaultResolver,
+}
+
+impl SodiumResolver {
+    pub fn new() -> Self {
+        SodiumResolver {
+            parent: DefaultResolver,
+        }
+    }
+}
+
+impl CryptoResolver for SodiumResolver {
+    fn resolve_rng(&self) -> Option<Box<Random + Send>> {
+        Some(Box::new(SodiumRandom::default()))
+    }
+
+    fn resolve_dh(&self, choice: &DHChoice) -> Option<Box<Dh + Send>> {
+        match *choice {
+            DHChoice::Curve25519 => Some(Box::new(SodiumDh25519::default())),
+            _ => self.parent.resolve_dh(choice),
+        }
+    }
+
+    fn resolve_hash(&self, choice: &HashChoice) -> Option<Box<Hash + Send>> {
+        match *choice {
+            HashChoice::SHA256 => Some(Box::new(SodiumSha256::default())),
+            _ => self.parent.resolve_hash(choice),
+        }
+    }
+
+    fn resolve_cipher(&self, choice: &CipherChoice) -> Option<Box<Cipher + Send>> {
+        match *choice {
+            CipherChoice::ChaChaPoly => Some(Box::new(SodiumChaChaPoly::default())),
+            _ => self.parent.resolve_cipher(choice),
+        }
+    }
+}
+
+// Random data generator.
+struct SodiumRandom;
+
+impl Default for SodiumRandom {
+    fn default() -> SodiumRandom {
+        SodiumRandom {}
+    }
+}
+
+impl Random for SodiumRandom {
+    fn fill_bytes(&mut self, out: &mut [u8]) {
+        let bytes: Vec<u8> = thread_rng().gen_iter::<u8>().take(out.len()).collect();
+        out.copy_from_slice(&bytes);
+    }
+}
+
+// Elliptic curve 25519.
+pub struct SodiumDh25519 {
+    privkey: sodium_curve25519::Scalar,
+    pubkey: sodium_curve25519::GroupElement,
+}
+
+impl Default for SodiumDh25519 {
```

Comment:
Why we need `default` implementation?

Reply:
Anyway using `[0;32]`( which is probably invalid `keypair`) is not a good practises as for me.

As i understand, `snow` init `DHProvider` with default value, and later sets the private key? Than we probably should make `pubkey`\`privkey` optional and `assert`s in methods is keys was really set.
@DarkEld3r ^

Reply:
I based my implementation on `crypto_wrapper` from `snow`. AFAICS it also [uses](https://github.com/mcginty/snow/blob/master/src/wrappers/crypto_wrapper.rs) zeros.

Does it make sense to create some random valid key if it's not really usable until user sets the key manually?

Reply:
>Does it make sense to create some random valid key if it's not really usable until user sets the key manually?

I don't like unspoken rules.
With external unstable library everything could break at any time, and if this (documented?) contract would change, than left it without `asserts` can lead a real debug hell in future.

May be it would good to add `assert_neq!(ZERO_KEY, key)` at method preludes?